feat(common): verify that priority images have corresponding preconnect links

AndrewKushnir opened the pull request:
This commit updates the `NgOptimizedImage` directive to add a logic to detect whether an image, marked with the "priority" attribute has corresponding `<link rel="preconnect">` tag in the `document.head`.


## PR Type
What kind of change does this PR introduce?

- [x] Feature


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
feat(common): verify that priority images have preconnect links

This commit updates the `NgOptimizedImage` directive to add a logic to detect whether an image, marked with the "priority" attribute has a corresponding `<link rel="preconnect">` tag in the `document.head` for better performance.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image.ts:
```diff
@@ -135,6 +130,112 @@ class LCPImageObserver implements OnDestroy {
+    // change detection cycle and query for the `<link rel="preload">` and
+    // `<link rel="preconnect">` once (vs doing that for each image once we
+    // come across one). This helps avoid unnecessary and repeated DOM lookups.
+    if (!this.onStable) {
```
Note to the reviewers: I plan to add some extra tests to make sure this logic works as expected.

pkozlowski-opensource commented:
@AndrewKushnir would you mind rebasing this PR? Kind of hard to review it in the current state.

AndrewKushnir commented:
@pkozlowski-opensource the PR is now rebased, thanks for the comment.

khempenius commented:
Summarizing offline discussion:
- I think it makes sense to hold off on checking for preload resource hints until Angular CLI can auto-insert them. Without automated tooling it could be difficult to ensure that the preload tags stay in sync with page content & if they were to get out of sync, this would be bad for performance.
- We will still verify the presence of preconnect resource hints - however this can be opted out on a per-loader basis. For example, sites that use different origins to serve the main document vs. images during development - but serve both from the same origin during production, would probably want to use this opt-out.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image.ts:
```diff
@@ -147,6 +142,112 @@ class LCPImageObserver implements OnDestroy {
+ * thus there is no `ngDevMode` use in the code.
+ */
+@Injectable({providedIn: 'root'})
+export class PreloadLinkChecker {
```
nit: could we move this class to a separate file? It is getting a bit crowded around here :-)

AndrewKushnir replied:
Yeah, I will move it in a separate file in a followup PR :)

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image.ts:
```diff
@@ -147,6 +142,112 @@ class LCPImageObserver implements OnDestroy {
+    const preconnect = new Set();
+    // Look for preload and preconnect links in one query
+    // and separate them into distinct sets later.
+    const selector = 'link[rel=preload],link[rel=preconnect]';
```
Open question: isn't the user supposed to manually add those `<link rel="preconnect">` tags manually in the index.html? What is the expected mechanism for the set of those tags to change? 

What I'm getting at is that if we expect (again, at least initially) to have a fixed / static set of tags maybe we don't need the whole machinery of detecting changes in the set of available `<link>` tags?

AndrewKushnir replied:
Good point. Based on the offline discussion, the directive would treat the list of `<link>`s as a static set.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image.ts:
```diff
@@ -147,6 +142,112 @@ class LCPImageObserver implements OnDestroy {
 
+/**
+ * Contains the logic to detect whether an image, marked with the "priority" attribute
+ * has corresponding `<link rel="preload">` or `<link rel="preconnect">` tag in the
```
Do we still want to verify both  `<link rel="preconnect">` and `<link rel="preload">`? 

Based on https://github.com/angular/angular/pull/46053#issuecomment-1138731062 it sounded like we want to verify the  `<link rel="preconnect">` only?

AndrewKushnir commented on packages/private/testing/src/render3.ts:
```diff
@@ -32,14 +31,49 @@ import {ɵresetJitOptions as resetJitOptions} from '@angular/core';
+ * @param blockFn function to wrap. The function can return promise or be `async`.
+ * @publicApi
+ */
+export function withHead<T extends Function>(html: string, blockFn: T): T {
```
Note to the reviewers: I will create a separate PR against the main Angular branches and land it. This PR would **not** contain this change, it's just there at this moment for completeness (to make things work).

AndrewKushnir replied:
The PR with those changes for the main Angular branches: https://github.com/angular/angular/pull/46250.

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -119,6 +112,76 @@ class LCPImageObserver implements OnDestroy {
+      console.warn(formatRuntimeError(
+          RuntimeErrorCode.PRIORITY_IMG_MISSING_PRECONNECT_TAG,
+          `${imgDirectiveDetails(rawSrc)} has detected that this image ` +
+              `contains the "priority" attribute, but don't have a corresponding ` +
```
```suggestion
              `contains the "priority" attribute, but doesn't have a corresponding ` +
```

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -119,6 +112,76 @@ class LCPImageObserver implements OnDestroy {
+          `${imgDirectiveDetails(rawSrc)} has detected that this image ` +
+              `contains the "priority" attribute, but don't have a corresponding ` +
+              `preconnect tag. Please add the following element into ` +
+              `the <head> of the document for better performance:\n` +
```
```suggestion
              `the <head> of the document to optimize loading of this image:\n` +
```

kara commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -495,6 +497,91 @@ describe('Image directive', () => {
+         expect(consoleWarnSpy.calls.count()).toBe(0);
+       }));
+
+    it('should log a warning if there is no matching preconnect link for a priority image',
```
Nit

```suggestion
    it('should log a warning if there is a preconnect, but it doesn't match the priority image',
```

kara commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -495,6 +497,91 @@ describe('Image directive', () => {
+                 'with the `rawSrc="a.png"`) has detected that this image contains the "priority" ' +
+                 'attribute, but don\'t have a corresponding preconnect tag. Please add the ' +
+                 'following element into the <head> of the document for better performance:' +
+                 '\n  <link rel="preconnect" href="http://localhost">');
```
Can we replace localhost with a more realistic situation, e.g. an image URL that has a real origin?

Also, seems like we should never be telling users to add a preconnect tag pointing to localhost? Should we have a special case for that (maybe in a follow-up)?

kara reviewed (commented):
Thanks for making all the changes! A few minor things below

AndrewKushnir committed:
feat(common): provide an ability to exclude origins from preconnect checks in NgOptimizedImage

This commit adds an extra logic to add an ability to exclude origins from preconnect checks in NgOptimizedImage by configuring the `PRECONNECT_CHECK_BLOCKLIST` multi-provider.

AndrewKushnir commented:
@kara @pkozlowski-opensource FYI this PR is updated and is ready for the final review. Thank you.

pkozlowski-opensource commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -495,6 +498,200 @@ describe('Image directive', () => {
+    describe('PRECONNECT_CHECK_BLOCKLIST token', () => {
+      it(`should allow passing host names`, withHead('', () => {
+           const providers = [
+             {provide: PRECONNECT_CHECK_BLOCKLIST, useValue: 'angular.io', multi: true},
```
Question: do we want it to be a multi-provider? What about an alternative where a user is providing a collection of host names?

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,133 @@
+
+  constructor(
+      @Inject(DOCUMENT) private doc: Document,
+      @Optional() @Inject(PRECONNECT_CHECK_BLOCKLIST) blocklist: Array<string|string[]>|null) {
```
nit question: since this class constructor is meant to run in the dev mode only, should be check and throw when it is not? We've got such code in `core` to make sure that bigger chunks of code don't end up in prod builds.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,133 @@
+  private pupulateBlocklist(origins: Array<string|string[]>) {
+    if (Array.isArray(origins)) {
+      deepForEach(origins, origin => {
+        // If an origin is an absolute URLs, extract the hostname only.
```
nit: extract this logic of the hostname extraction to a separate, named function?

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  check(rewrittenSrc: string, rawSrc: string) {
+    if (!this.window) return;
```
question: not sure why do we have this check? Could you explain / add comment?

AndrewKushnir replied:
We rely on the `window` presence (`window.location`) to properly construct the `URL` instance (for relative URLs specifically), so if the window is not present (in some non-browser environments?) - we can't proceed, thus early exit.

pkozlowski-opensource reviewed (commented):
@AndrewKushnir I did another pass on this PR and left a couple of nit / question comments but things LGTM overall. 

The only open question I've got is the usage of a multiprovider for the preconnect blocklist. Since the application author / "integrator" is in charge of adding the procennect links it should be, IMO, an app responsibility to configure loaders, add preconnect tags etc. 

In my head other modules (ex. 3rd party loaded from NPM) should have no say in this blocklist. Let's discuss in person as we've got a dedicated meeting today.

AndrewKushnir committed:
fixup! feat(common): provide an ability to exclude origins from preconnect checks in NgOptimizedImage

AndrewKushnir commented:
Quick update: after an offline discussion, the `PRECONNECT_CHECK_BLOCKLIST` token remains a multi-provider one to take into account a situation when loaders are provided on a component level (thus there is a potential for multiple host names that we'd need to check). The rest of the feedback is addressed in this fixup commit: https://github.com/angular/angular/pull/46053/commits/d7a3b91d2b88543b8681757d6c4dfaa71790d048. @pkozlowski-opensource could you please take another look?

pkozlowski-opensource reviewed (approved):
LGTM
